### PR TITLE
tests: use stages for the travis execution

### LINKED ISSUE #694: Travis coverage job should not run linting
The way Travis was [set up](https://github.com/OpenZeppelin/zeppelin-solidity/blob/ff9e9c4d8587a2504b04a39f2e81f382de1e0fc0/.travis.yml) in #369 allows to run coverage and tests in parallel. However, the coverage job is also running the linters, redundantly.

### PR DESCRIPTION
**Pull request by elopio**
Fixes #694

# 🚀 Description

Use travis stages for better control of the things to run in parallel.

- [x] 📘 I've reviewed the [OpenZeppelin Contributor Guidelines](../blob/master/CONTRIBUTING.md)
- [x] ✅ I've added tests where applicable to test my new functionality.
- [x] 📖 I've made sure that my contracts are well-documented.
- [x] 🎨 I've run the JS/Solidity linters and fixed any issues (`npm run lint:all:fix`).

### CODE REVIEW

**Comment by elopio** [comment]
I'm not super sure about fast finish and allow failures. But it's late and travis is grumpy and not taking this PR. So, WIP, please don't merge it yet.

**Comment by elopio** [comment]
Okay, ready for review.

File: .travis.yml [hunk 1]
```diff
@@ -4,24 +4,26 @@ group: beta
...
-  - npm run test
+  include:
+    - stage: static
+      script: npm run lint
+    - stage: unit
+      script: npm run test
+    - script: npm run test
+      env: SOLIDITY_COVERAGE=true
+    - script: npm run test
+      env: SOLC_NIGHTLY=true
```
Comment (nventuro) [inline .travis.yml]: I'd add a short comment explaining we're running 3 jobs in the `unit` stage, since Travis' syntax for this is so obscure.

Comment (nventuro) [inline .travis.yml]: I'd add a short comment explaining we're running 3 jobs in the unit stage, since Travis' syntax for this is so obscure.

Comment (elopio) [reply .travis.yml]: Thanks for the review @nventuro!
I've added comments, please take another look.

**Comment by nventuro** [comment]
Apparently Travis supports explicitly indicating the stage for each job: I think this would make the intent clearer.

This config:
```
  include:
    - stage: static
      script: npm run lint
    - stage: unit
      script: npm run test
    - stage: unit
      script: npm run test
      env: SOLIDITY_COVERAGE=true
    - stage: unit
      script: npm run test
      env: SOLC_NIGHTLY=true
```
produces the [same result](https://travis-ci.org/nventuro/openzeppelin-solidity).

**Comment by elopio** [comment]
@nventuro updated.

File: package.json [hunk 2]
```diff
@@ -8,12 +8,12 @@
   ],
   "scripts": {
     "test": "scripts/test.sh",
-    "lint": "eslint .",
-    "lint:fix": "eslint . --fix",
+    "lint:js": "eslint .",
```
Comment (frangio) [inline package.json]: Yay! This was in my mind too.

File: .travis.yml [hunk 3]
```diff
@@ -4,24 +4,37 @@ group: beta
...
+    - stage: unit
+      script: npm run test
+    - stage: unit
+      script: npm run test
+      env: SOLIDITY_COVERAGE=true
+    - stage: unit
+      script: npm run test
+      env: SOLC_NIGHTLY=true
+    # solidity and javascript style tests.
+    - stage: static
```
Comment (frangio) [inline .travis.yml]: I don't think "Static" is a very clear name. Have you considered "Linting"?

File: .travis.yml [hunk 4]
```diff
@@ -4,24 +4,37 @@ group: beta
...
-  - npm run lint
-  - npm run lint:sol
-  - npm run test
+  include:
+    # Run the unit test suite three times in parallel.
+    # The first one gets results faster and is the only one required to pass.
+    # The second one generates the coverage report.
+    # The third one is to keep us informed about possible issues with the
+    # upcoming solidity release.
+    - stage: unit
```
Comment (frangio) [inline .travis.yml]: How about "Unit tests" for the stage name?

File: package.json (hunk 2 above)
Comment (frangio) [reply package.json]: If we put this PR on hold we should still do this now.

**Comment by frangio** [comment]
@elopio Can you explain a bit further the motivation behind this? Were you looking to put linting in a separate job? If so, what is the problem with it running in parallel with everything else?

File: .travis.yml (hunk 4 above)
Comment (elopio) [reply .travis.yml]: sounds good.

File: .travis.yml (hunk 3 above)
Comment (elopio) [reply .travis.yml]: wfm.

**Comment by nventuro** [comment]
Fast finish [does work](https://github.com/travis-ci/beta-features/issues/28#issuecomment-387182805), but [stages won't advance](https://github.com/travis-ci/travis-ci/issues/9677) until all jobs (even those that are allowed to fail) finish. So we could have the best of both worlds, by putting the linter stage first. Given our decision of not treating aesthetic linter errors as build errors, this is fine by me.

**Comment by frangio** [comment]
It's still not clear to me if having the unit tests as the last stage will result in `fast_finish` actually causing the entire build to fail once the tests fail in the non-coverage non-nightly job.

**Comment by nventuro** [comment]
I'll do some tests on this and come back with more information.

**Comment by frangio** [comment]
It seems like linting will still not run until all jobs including coverage have finished.

**Comment by elopio** [comment]
Wow, and the diff is now looking super ugly.
 
@frangio we could run linting in parallel with any of the stages, but now I'm not sure what you are looking for :D
@nventuro convinced you make the linter fail on error, so why not move it back as the first step of the test run?

**Comment by nventuro** [comment]
@elopio we'd like for:

a) linting errors to cause a build error, but still allow for the test suite to run (i.e. don't run linting first and abort the build on error)
b) linting errors to not wait until the coverage test has finished, since that can take very long

**Comment by elopio** [comment]
@nventuro So, all of them running in parallel. That is, no stages.
But, I still prefer the stages syntax over the matrix, and #1186 will look a lot better as a conditional stage.
I will make a simple change to run all of them in parallel.

**Comment by elopio** [comment]
@nventuro @frangio please take another look. But we could argue here endlessly about the details, I would prefer to land it and do small tweaks when we have a reason to complain about the way the tests are executed.

**Pull request merged**